src/common/promises/CancelToken.ts

```diff
@@ -0,0 +1,272 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the MIT license found in the
+ * LICENSE file in the root of this projects source tree.
+ */
+
+import { AnyFunction, BasicEvent } from "./types";
+import { defer, DeferredResult } from "./defer";
+
+import { $$toStringTag } from "./symbols";
+import { BaseError } from "../errors";
+import { isPromise } from "./isPromise";
+import { noop } from "./noop";
+
+export type CancelAction = (message: string | Cancel) => void;
+
+// eslint-disable-next-line @typescript-eslint/typedef
+const cancelTokenTag = "CancelToken" as const;
+
+const InternalActionGetter: (action: CancelAction) => void = (_action: CancelAction) => {
+    // noop
+};
+
+type InternalHandler = AnyFunction & { listener: AnyFunction | { handleEvent: AnyFunction } };
+
+export class Cancel extends BaseError {
+    // private readonly message: string;
+
+    constructor(message: string = "this action has been canceled") {
+        super(message);
+    }
+}
+
+export class CancelToken {
```

Comment:
A more standard term seems to be `CancellationToken`, but I don't feel strongly about this. 

Are there existing cancellation interfaces in the vscode APIs? Would it make sense to use them here?

Reply:
rename CancelToken into CancellationToken;
rename CancelSource into CancellationTokenSource;
and also implement vsc's CancellationToken for future integration.

I was thinking, we gonna have to adapt nodejs abort onAbort signal, or even other cancellation token from its source, like vsc cancellation tokens.
Thus, I was thinking it would be safer to have our own cancellation implementation as a buffer layer where we could combine or adapt to any when we needed. 😬